src/censure.jl

```diff
@@ -0,0 +1,223 @@
+abstract BiFilter
+
+type BoxFilter <: BiFilter
+
+	scale :: Integer
+	in_length :: Integer
+	out_length :: Integer
+	in_area :: Float64
+	out_area :: Float64
+	in_weight :: Float64
+	out_weight :: Float64
+
+end
+
+type OctagonFilter <: BiFilter
+	
+	m_out :: Integer
+	m_in :: Integer
+	n_out :: Integer
+	n_in :: Integer
+	in_area :: Float64
+	out_area :: Float64
+	in_weight :: Float64
+	out_weight :: Float64
+
+end
+
+type CENSURE <: Detector
+
+	smallest :: Integer
+	largest :: Integer
+	filter_type :: Type
+	filter_stack :: Array
+	responseThreshold :: Number
+	lineThreshold :: Number
+
+end
+
+function OctagonFilter(mo, mi, no, ni)
+	OF = OctagonFilter(mo, mi, no, ni, 0.0, 0.0, 0.0, 0.0)
+	OF.out_area = OF.m_out ^ 2 + 2 * OF.n_out ^ 2 + 4 * OF.m_out * OF.n_out
+	OF.in_area = OF.m_in ^ 2 + 2 * OF.n_in ^ 2 + 4 * OF.m_in * OF.n_in
+	OF.out_weight = 1.0 / (OF.out_area - OF.in_area)
+	OF.in_weight = 1.0 / OF.in_area
+	OF
+end
+
+function BoxFilter(s)
+	BF = BoxFilter(s, 2 * s + 1, 4 * s + 1, (2 * s + 1) ^ 2, (4 * s + 1) ^ 2, 0.0, 0.0)
+	BF.in_weight = 1.0 / BF.in_area
+	BF.out_weight = 1.0 / (BF.out_area - BF.in_area)
+	BF
+end
+
+OctagonFilter_Kernels = [
+						[5, 3, 2, 0],
+						[5, 3, 3, 1],
+						[7, 3, 3, 2],
+						[9, 5, 4, 2],
+						[9, 5, 7, 3],
+						[13, 5, 7, 4],
+						[15, 5, 10, 5]
+						]
+
+BoxFilter_Kernels = [1, 2, 3, 4, 5, 6, 7]
+
+Kernels = Dict(BoxFilter => BoxFilter_Kernels, OctagonFilter => OctagonFilter_Kernels)
+
+function getFilterStack(filter_type::Type, smallest::Integer, largest::Integer)
+	k = Kernels[filter_type]
+	filter_stack = map(f -> filter_type(f...), k[smallest : largest])
+end
+
+getIntegralImage(img, filter_type::BoxFilter) = integral_image(img)
+
+function getIntegralImage(img, filter_type::OctagonFilter)
+	img_shape = size(img)
```

Comment:
This function computes the integral and slant-integral images (right_slant holds the sum of pixels to the top and left as well as the pixels to the top and right till an angle of 45 degrees whereas left_slant holds the sum of pixels to the top and left till an angle of 45 degrees.)

Here the functions go in a row major fashion. I wanted to use CartesianRange but it goes in a coloumn major fashion which is the way julia stores arrays. Does the ```Image``` type store data in a row major fashion? I was a bit confused after going through the documentation.

I could transpose the image matrix and then take the integral images which will be faster if it is in a coloumn major fashion. But transposing just gives a view of the matrix right? Does it actually transpose the matrix in memory?